### PR TITLE
iostream: Add with_output_stream helper

### PR DESCRIPTION
**Pull request by xemul**
The wrapper get a stream and function and makes sure the stream is flushed and closed at the end regardless of any exceptions that may pop up along the way.

Test included.

### CODE REVIEW

**Review by bhalevy — APPROVED** [review]

**Comment by xemul** [comment]
@scylladb/seastar-maint , review ping

File: include/seastar/core/iostream.hh [hunk 1]
```diff
@@ -439,6 +439,26 @@ private:
...
+///
+/// \param out A stream to work on
+/// \param func A function that uses the stream
+/// \returns the future returned by \c func, or an exceptional future if either flushing
+/// or closing the \c out failed.
+template <typename Func, typename CharType, typename Result = typename internal::future_result<Func, output_stream<CharType>&>::future_type>
+SEASTAR_CONCEPT( requires std::invocable<Func, output_stream<CharType>&> && std::is_nothrow_move_constructible_v<Func> )
+Result with_output_stream(output_stream<CharType>&& out, Func func) noexcept {
+    static_assert(std::is_nothrow_move_constructible_v<Func>, "Func's move constructor must not throw");
+    return do_with(std::move(out), [func = std::move(func)] (output_stream<CharType>& out) mutable {
```
Comment (bhalevy) [inline include/seastar/core/iostream.hh:454]: It makes sense here to accept `Func&& func` and here to capture it by `func = std::forward<Func>(func)` and save a move.

**Comment by xemul** [comment]
upd:
- use rvalure ref for Func argument

**Review by bhalevy — APPROVED** [review]
LGTM, thanks

File: include/seastar/core/iostream.hh [hunk 2]
```diff
@@ -439,6 +439,27 @@ private:
...
+/// \param func A function that uses the stream
+/// \returns the future returned by \c func, or an exceptional future if either flushing
+/// or closing the \c out failed.
+template <typename Func, typename CharType>
+SEASTAR_CONCEPT( requires std::invocable<Func, output_stream<CharType>&> && std::is_nothrow_move_constructible_v<Func> )
+auto with_output_stream(output_stream<CharType>&& out, Func func) noexcept {
+    static_assert(std::is_nothrow_move_constructible_v<Func>, "Func's move constructor must not throw");
+    return do_with(std::move(out), [func = std::move(func)] (output_stream<CharType>& out) mutable {
+        return futurize_invoke(func, out).finally([&out] {
+            return out.flush().finally([&out] {
```
[the remark this answers is not in the document]
Comment (avikivity) [reply include/seastar/core/iostream.hh]: Yes, throwing from close() is a very bad idea as it's called in error recovery paths.

flush() is like commit, while close() is like a destructor.

File: include/seastar/core/iostream.hh [hunk 3]
```diff
@@ -439,6 +439,26 @@ private:
...
+///
+/// The \c out rvalue is moved into stable location and is passed to \c func.
+///
+/// \param out A stream to work on
+/// \param func A function that uses the stream
+/// \returns the future returned by \c func, or an exceptional future if either flushing
+/// or closing the \c out failed.
+template <typename Func, typename CharType, typename Result = typename internal::future_result<Func, output_stream<CharType>&>::future_type>
+SEASTAR_CONCEPT( requires std::invocable<Func, output_stream<CharType>&> && std::is_nothrow_move_constructible_v<Func> )
+Result with_output_stream(output_stream<CharType>&& out, Func&& func) noexcept {
```
Comment (avikivity) [inline include/seastar/core/iostream.hh:452]: Isn't this with_closeable() now?

Comment (bhalevy) [reply include/seastar/core/iostream.hh:452]: good catch.  it should be

Comment (xemul) [reply include/seastar/core/iostream.hh:452]: :fire:

**Pull request closed without merging**